src/spectralanalysis.jl

```diff
@@ -1,7 +1,27 @@
 using Arpack
+import KrylovKit: eigsolve
 
 const nonhermitian_warning = "The given operator is not hermitian. If this is due to a numerical error make the operator hermitian first by calculating (x+dagger(x))/2 first."
 
+mutable struct DiagStrategy{DT}
+    n::Int
+    v0::Union{Nothing, AbstractVector}
+end
+DiagStrategy{T}(n::Int) where T = DiagStrategy{T}(n, nothing)
+const LapackDiag = DiagStrategy{:lapack}
+const ArpackDiag = DiagStrategy{:arpack}
+const KrylovDiag = DiagStrategy{:krylov}
```

Comment:
could you consider structuring this a bit differently:

```
abstract type DiagStrategy end
type LapackDiag{T} <: DiagStrategy
    n::Int
    v0::T
end
```

My main reasons for this suggestion:

- It is a bit closer in style to how SciML treats their solver types
- It is a bit closer in style to Holy Traits https://invenia.github.io/blog/2019/11/06/julialang-features-part-2/
- It parameterizes `v0` instead of leaving it of abstract type, which avoids boxing (but to be fair, you can do that to your version either way)
- It is not mutable anymore

Please feel free to voice dissent to this suggestion (as always!)

Reply:
I was wrong on the last one: it still needs to be mutable because you might change `n` as I see below

Reply:
I agree, this is a better way. Initially I used this structure to make custom subtyping way easier, so that LapackDiag, ArpackDiag and KrylovDiag inherited the same internal structure. 
Probably a good way to avoid the `DiagStrategy` struct being mutable is making the `DiagStrategy` constructor process the keyword arguments from the `eigenstates` method. This explanation is probably kinda messy, will clear it out in the nearest commit.